impeller/entity/contents/color_source_contents.h

```diff
@@ -112,80 +113,83 @@ class ColorSourceContents : public Contents {
           ContentContextOptions)>;
 
   template <typename VertexShaderT>
-  bool DrawGeometry(GeometryResult geometry_result,
-                    const ContentContext& renderer,
+  bool DrawGeometry(const ContentContext& renderer,
                     const Entity& entity,
                     RenderPass& pass,
                     const PipelineBuilderCallback& pipeline_callback,
                     typename VertexShaderT::FrameInfo frame_info,
-                    const BindFragmentCallback& bind_fragment_callback) const {
+                    const BindFragmentCallback& bind_fragment_callback,
+                    bool enable_uvs = false,
+                    Rect texture_coverage = {},
+                    const Matrix& effect_transform = {}) const {
     auto options = OptionsFromPassAndEntity(pass, entity);
+
+    GeometryResult::Mode geometry_mode = GetGeometry()->GetResultMode();
+    Geometry& geometry = *GetGeometry();
+
+    const bool is_stencil_then_cover =
+        geometry_mode == GeometryResult::Mode::kNonZero ||
+        geometry_mode == GeometryResult::Mode::kEvenOdd;
+    if (is_stencil_then_cover) {
+      pass.SetStencilReference(0);
+
+      /// Stencil preparation draw.
+
+      GeometryResult stencil_geometry_result =
+          GetGeometry()->GetPositionBuffer(renderer, entity, pass);
+      pass.SetVertexBuffer(std::move(stencil_geometry_result.vertex_buffer));
+      options.primitive_type = stencil_geometry_result.type;
+
+      options.blend_mode = BlendMode::kDestination;
+      switch (stencil_geometry_result.mode) {
+        case GeometryResult::Mode::kNonZero:
+          pass.SetCommandLabel("Stencil preparation (NonZero)");
+          options.stencil_mode =
+              ContentContextOptions::StencilMode::kStencilNonZeroFill;
+          break;
+        case GeometryResult::Mode::kEvenOdd:
+          pass.SetCommandLabel("Stencil preparation (EvenOdd)");
+          options.stencil_mode =
+              ContentContextOptions::StencilMode::kStencilEvenOddFill;
+          break;
+        default:
+          FML_UNREACHABLE();
+      }
+      pass.SetPipeline(renderer.GetClipPipeline(options));
```

Comment:
I had to rework this StC logic a bit because it wasn't handling Positions+UVs right. I was inadvertently stenciling with position+UV geometry (bad) and then covering using a position-only rectangle (also bad).

This is fixed now. When we need to use StC, we only generate positions for the stencil prep draw and then draw the cover rectangle with or without UVs as necessary.

Reply:
I have a preference towards separating out this rework and the flag-flip of the stencil-then-cover into separate PRs.

Reply:
Agreed, I pulled out these changes into this PR: https://github.com/flutter/engine/pull/50900